
<h1 align="center"> Battle of the Neighbourhoods - Toronto </h1>

<h2 align="center"> Felipe Meza </h2>

In [2]:
import pandas as pd

# Part 1

### Importing Data

In [3]:
import requests

In [4]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Since we are receiveing the message "Response 200  that means that connection to the page is ok.

In [5]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [6]:
len(wiki_data), type(wiki_data)

(3, list)

We drop the tables and we just keep the first one. 

In [7]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough (not assigned)

In [8]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

We group based on Postal Code

In [9]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Validating the number of records where Neighbourhood is "Not assigned"

In [10]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [11]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

Dropping "index"

In [12]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [13]:
df.shape

(103, 3)

Number of rows of the dataframe: 103 rows...and 3 columns

# Part 2

geocoder installation 

In [14]:
pip install geocoder

     |████████████████████████████████| 98 kB 9.0 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [15]:
import geocoder # import geocoder

```python
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
```

Importing the CSV file from the URL

In [16]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [17]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Taking avantage of the same dimentions, we can combine the data.

First, we check the column types of the DF's.

In [18]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [19]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [20]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [21]:
combined_data.shape

(103, 5)

We get 103 rows as expected.

# Part 3

We learn in the last lab to cluster the neighbourhood of NYC, now we will do somethong similar but for Toronto, using similar categories and Kmeans.

In [22]:
from geopy.geocoders import Nominatim 

In [23]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


folium installation 

In [26]:
pip install folium

     |████████████████████████████████| 94 kB 5.6 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


Map of Toronto

In [27]:
import folium

In [28]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

My Foursquare API credentials:

In [29]:
CLIENT_ID = 'BV43R1M33NVDNFH4QV3WRO3VNRXEIX5UCSOC0L1RCTGBYUKF' # your Foursquare ID
CLIENT_SECRET = '5QVHIAXWSSMUOA4QR2XJYXF3EDHQF0QUM1DK0RZMVYD1QDDG' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: BV43R1M33NVDNFH4QV3WRO3VNRXEIX5UCSOC0L1RCTGBYUKF
CLIENT_SECRET:5QVHIAXWSSMUOA4QR2XJYXF3EDHQF0QUM1DK0RZMVYD1QDDG


Function to obtain the venues categories in Toronto

In [30]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

We collect the venues in Toronto for each Neighbourhood

In [31]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [32]:
venues_in_toronto.shape

(1328, 5)

In [33]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

In [34]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   
1316  Mimico NW, The Queensway West, South of Bloor,...   
1317  Mimico NW, The Queensway West, South of Bloor,...   
1318  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1314               43.628841               -79.520999              Wingporium   
1315               43.628841               -79.520999        South St. Burger   
1316               43.628841               -79.520999               Dollarama   
1317               43.628841               -79.520999          Healthy Planet   
1318               43.628841               -79.520999                7-Eleven   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1314            Wings Joint  
1315           Burger Joint  
1316         Discount Store  
1317        Supplement Shop  
1318      Convenience Store  

[412 rows x 5 columns]

We have 405 records for each neighbourhood.

What is the maximum venue categories?

In [35]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Wine Bar                                     Little Portugal, Trinity   
Wine Shop                                Dufferin, Dovercourt Village   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750072               -79.295849   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Wine Bar                         43.653206               -79.400049   
Wine Shop                        43.669005               -79.442259   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.733283               -79.419750   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Wine Bar                            Paris Paris Bar  
Wine Shop                                    Macedo  
Wings Joint                              Wingporium  
Women's Store                         Want Boutique  
Yoga Studio                               YogaSpace  

[235 rows x 4 columns]

It seems to be 235 different types of Venue Categories.

### One Hot encoding the venue Categories

In [36]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1323                  0               0        0                   0   
1324                  0               0        0                   0   
1325                  0               0        0                   0   
1326                  0               0        0                   0   
1327                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1323             0               0                0                 0   
1324             0               0                0                 0   
1325             0               0                0                 0   
1326             0               0                0                 0   
1327             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Train Station  \
0                       0             0  ...              0   
1                       0             0  ...              0   
2                       0             0  ...              0   
3                       0             0  ...              0   
4                       0             0  ...              0   
...                   ...           ...  ...            ...   
1323                    0             0  ...              0   
1324                    0             0  ...              0   
1325                    0             0  ...              0   
1326                    0             0  ...              0   
1327                    0             0  ...              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1323                              0                 0                      0   
1324                              0                 0                      0   
1325                              0                 0                      0   
1326                              0                 0                      0   
1327                              0                 0                      0   

      Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                   0         0          0            0              0   
1                   0         0          0            0              0   
2                   0         0          0            0              0   
3                   0         0          0            0              0   
4                   0         0          0            0              0   
...               ...       ...        ...          ...            ...   
1323                0       

We add the neighbourhood to the encoded DF

In [37]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2  Victoria Village                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  \
0                 0                    0  ...              0   
1                 0                    0  ...              0   
2                 0                    0  ...              0   
3                 0                    0  ...              0   
4                 0                    0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                0         0          0            0              0   
1                0         0          0            0              0   
2                0         0          0            0              0   
3                0         0          0            0              0   
4                0         0          0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 236 columns]

Grouping the Neighbourhoods and calculating the mean venue categories in each Neighbourhood 

In [38]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  Train Station  \
0              0.0               0.0                 0.00  ...            0.0   
1              0.0               0.0                 0.00  ...            0.0   
2              0.0               0.0                 0.00  ...            0.0   
3              0.0               0.0                 0.00  ...            0.0   
4              0.0               0.0                 0.04  ...            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                            0.0               0.0                    0.0   
1                            0.0               0.0                    0.0   
2                            0.0               0.0                    0.0   
3                            0.0               0.0                    0.0   
4                            0.0               0.0                    0.0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0              0.0       0.0        0.0          0.0           0.00   
1              0.0       0.0        0.0          0.0           0.00   
2              0.0       0.0        0.0          0.0           0.00   
3              0.0       0.0        0.0          0.0           0.00   
4              0.0       0.0        0.0          0.0           0.04   

   Yoga Studio  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  

[5 rows x 236 columns]

What are the top most common venue categories?

In [55]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [40]:
import numpy as np

Since we have too many venue categories, lets use the top 10 to cluster the neighbourhoods

In [56]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

# return_most_common_venues
for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village         Chinese Restaurant   
4                Bedford Park, Lawrence Manor East                Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Clothing Store        Breakfast Spot   
1           Coffee Shop                   Pub        Sandwich Place   
2                  Bank         Deli / Bodega           Supermarket   
3                  Bank   Japanese Restaurant                  Café   
4        Sandwich Place    Italian Restaurant         Grocery Store   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0          Dance Studio             Drugstore                   Donut Shop   
1              Pharmacy                   Gym                  Curling Ice   
2      Sushi Restaurant            Restaurant                        Diner   
3           Yoga Studio         Deli / Bodega  Eastern European Restaurant   
4              Pharmacy               Butcher                         Café   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Dog Run   Distribution Center         Discount Store  
1               Dance Studio         Deli / Bodega       Department Store  
2  Middle Eastern Restaurant        Sandwich Place      Mobile Phone Shop  
3                  Drugstore            Donut Shop                Dog Run  
4                        Pub  Fast Food Restaurant       Sushi Restaurant

Cluster of our Neighbourhoods

In [42]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [43]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Labelling of our model

In [44]:
kmeans.labels_[0:100]

array([2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2, 2,
       2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2,
       2, 2, 2, 2, 2, 2, 4, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2,
       1, 4, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 0,
       1, 2, 1, 2, 0, 2, 2, 1], dtype=int32)

Adding the clustering Label column to the top 10 common venue categories

In [45]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare plotting

In [46]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             1.0     Food & Drink Shop   
1  43.725882 -79.315572             2.0           Coffee Shop   
2  43.654260 -79.360636             2.0           Coffee Shop   
3  43.718518 -79.464763             2.0        Clothing Store   
4  43.662301 -79.389494             2.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue       4th Most Common Venue  \
0                  Park           Yoga Studio                   Drugstore   
1     French Restaurant           Pizza Place  Financial or Legal Service   
2                  Park               Theater                      Bakery   
3     Accessories Store              Boutique      Furniture / Home Store   
4      Sushi Restaurant   Fried Chicken Joint                         Gym   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0            Donut Shop               Dog Run    Distribution Center   
1          Hockey Arena          Intersection  Portuguese Restaurant   
2                  Café        Breakfast Spot            Yoga Studio   
3           Event Space    Miscellaneous Shop             Shoe Store   
4                   Bar              Beer Bar          Smoothie Shop   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0        Discount Store                  Diner     Dim Sum Restaurant  
1          Dance Studio          Deli / Bodega       Department Store  
2          Dessert Shop                    Spa            Event Space  
3           Coffee Shop  Vietnamese Restaurant        Airport Service  
4        Sandwich Place           Burger Joint          Burrito Place

Droping NaN's values:

In [47]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [48]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [49]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

We have identified the 10 most common venues per place...If for example, we are looking to do busssiness in this city, that will give us a light, about the potential of our bessiness idea, based on that info.